src/common/hashtable.c

```diff
@@ -0,0 +1,351 @@
+        }
+    }
+
+    ht->base_size = new_ht->base_size;
+    ht->count = new_ht->count;
+
+    // To delete new_ht, we give it ht's size and items 
+    const int tmp_size = ht->size;
+    ht->size = new_ht->size;
+    new_ht->size = tmp_size;
+
+    onf_ht_item** tmp_items = ht->items;
+    ht->items = new_ht->items;
+    new_ht->items = tmp_items;
+
+    onf_ht_del_hash_table(new_ht);
+    return SUCCESS;
+}
+
+/*****************************************
+Name:  onf_ht_resize_up
+brief: resize up 
+return: 
+*****************************************/
+int onf_ht_resize_up(onf_ht_hash_table* ht) {
+    const int new_size = ht->base_size * 2;
+    return onf_ht_resize(ht, new_size);
+}
+
+/*****************************************
+Name:  onf_ht_resize_down
+brief: resize down
+return: 
+*****************************************/
+int onf_ht_resize_down(onf_ht_hash_table* ht) {
+    const int new_size = ht->base_size / 2;
+    return onf_ht_resize(ht, new_size);
+}
+
+/*****************************************
+Name:  onf_ht_del_item
+brief: delete hash item
+return: 
+*****************************************/
+void onf_ht_del_item(onf_ht_item* i) {
+    free(i->key);
+    free(i->value);
+    free(i);
+}
+
+/*****************************************
+Name:  onf_ht_del_hash_table
+brief: delete hash table
+return: 
+*****************************************/
+void onf_ht_del_hash_table(onf_ht_hash_table* ht) {
+    for (int i = 0; i < ht->size; i++) {
+        onf_ht_item* item = ht->items[i];
+        if (item != NULL) {
+            onf_ht_del_item(item);
+        }
+    }
+    free(ht->items);
+    free(ht);
+}
+
+/*****************************************
+Name:  onf_ht_hash
+brief: delete hash table
+return: 
+*****************************************/
+int onf_ht_hash(
+      const char* s, int len_s, const int a, const int m) 
+{
+    long hash = 0;
+    for (int i = 0; i < len_s; i++) {
+        hash += (long)pow(a, len_s - (i+1)) * s[i];
+        hash = hash % m;
+    }
+    return (int)hash;
+}
+
+/*****************************************
+Name:  onf_ht_get_hash
+brief: get hash table
+return: 
+*****************************************/
+int onf_ht_get_hash(
+    const char* s, int key_len, 
+    const int num_buckets, const int attempt) 
+{
+    const int hash_a = onf_ht_hash(
+                        s, key_len, ONF_HT_PRIME_1, num_buckets);
+    const int hash_b = onf_ht_hash(
+                        s, key_len, ONF_HT_PRIME_2, num_buckets);
+    return (hash_a + (attempt * (hash_b + 1))) % num_buckets;
+}
+
+/*****************************************
+Name:  onf_ht_insert
+brief: insert into hash table
+return: 
+*****************************************/
+int onf_ht_insert(
+      onf_ht_hash_table* ht, int key_len,  
+      int val_len, const char* key, const char* value) 
+{
+    const int load = (ht->count * 100) / ht->size;
+    if (load > 70) {
+        if(SUCCESS != onf_ht_resize_up(ht))
+	{
+	    return E_FAIL;
```

Comment:
IF resize fails then should we continue to insert the element ?

Reply:
the allocation of memory has failed during resize. So, the application is not in a stable state. So better to fail.